ios/CodePush/CodePush.m

```diff
@@ -488,44 +491,43 @@ - (void)applicationWillResignActive
         [mutableUpdatePackage setValue:[CodePushUpdateUtils modifiedDateStringOfFileAtURL:binaryBundleURL]
                                 forKey:BinaryBundleDateKey];
     }
-
+    
+    if (notifyProgress) {
+        [self setupFrameObserverForDownloadProgress];
+    }
+    
     [CodePushPackage
         downloadPackage:mutableUpdatePackage
         expectedBundleFileName:[bundleResourceName stringByAppendingPathExtension:bundleResourceExtension]
+        operationQueue:_methodQueue
         // The download is progressing forward
         progressCallback:^(long long expectedContentLength, long long receivedContentLength) {
-            dispatch_async(_methodQueue, ^{
-                // Notify the script-side about the progress
-                [self.bridge.eventDispatcher
-                    sendDeviceEventWithName:@"CodePushDownloadProgress"
-                    body:@{
-                            @"totalBytes":[NSNumber numberWithLongLong:expectedContentLength],
-                            @"receivedBytes":[NSNumber numberWithLongLong:receivedContentLength]
-                          }];
-            });
+            [self updateDownloadProgressForNextFrame:expectedContentLength
+                               receivedContentLength:receivedContentLength];
+            // If the download is completed, stop observing frame updates and synchronously send the last event.
+            if (expectedContentLength == receivedContentLength) {
```

Comment:
Would it be cleaner is `updateDownloadProgressForNextFrame` did this check? That way, this block simply requests that the progress be updated and doesn't know or care about how it's been scheduled/throttled.